### PR TITLE
add read-only cached_messages property to Client

### PR DESCRIPTION
**Pull request by khazhyk**
### Summary

For those of us who want access to this sweet trove of zero hop messages

### Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/utils.py [hunk 1]
```diff
@@ -78,6 +79,17 @@ def decorator(func):
         return CachedSlotProperty(name, func)
     return decorator
 
+class SequenceProxy(collections.abc.Sequence):
```
Comment (Rapptz) [inline discord/utils.py]: This is missing the following to satisfy the interface:

1. `__iter__`
2. `__reversed__`
3. `__contains__`

File: discord/client.py [hunk 2]
```diff
@@ -230,6 +230,11 @@ def emojis(self):
         """List[:class:`Emoji`]: The emojis that the connected client has."""
         return self._connection.emojis
 
+    @property
+    def cached_messages(self):
+        """SequenceProxy[:class:`Message`]: Read-only list of messages the connected client has cached."""
```
Comment (Rapptz) [inline discord/client.py]: Needs a `..versionadded::` string.

File: discord/utils.py (hunk 1 above)
Comment (khazhyk) [reply discord/utils.py]: Those are implemented by the abc

**Comment by NCPlayz** [comment]
In [migrating.rst](https://discordpy.readthedocs.io/en/latest/migrating.html#models-are-stateful), it is mentioned that `Client.messages` was removed, but would it be feasible to write that it has returned in some form?

**Comment by khazhyk** [comment]
Updated migrating.rst. I don't have an rst renderer but it looks right...

File: discord/utils.py [hunk 3]
```diff
@@ -78,6 +79,32 @@ def decorator(func):
...
+    def __contains__(self, item):
+        return item in self.__proxied
+
+    def __iter__(self):
+        return iter(self.__proxied)
+
+    def __reversed__(self):
+        return reversed(self.__proxied)
+
+    def index(self, value, *args, **kwargs):
```
Comment (bmintz) [inline discord/utils.py]: Index takes no kwargs.

Comment (Rapptz) [reply discord/utils.py]: This is a forwarding function so it's fine.

File: docs/migrating.rst [hunk 4]
```diff
@@ -206,8 +206,6 @@ A list of these changes is enumerated below.
 +---------------------------------------+------------------------------------------------------------------------------+
 | ``Client.wait_until_login``           | Removed                                                                      |
 +---------------------------------------+------------------------------------------------------------------------------+
-| ``Client.messages``                   | Removed                                                                      |
```
Comment (Rapptz) [inline docs/migrating.rst]: Why not just point to the new attribute in the table instead of adding it somewhere else?

Comment (khazhyk) [reply docs/migrating.rst]: This changed didn't feel like it fit under the "models are stateful" heading so I moved it to the misc section

**Pull request closed without merging**